fix: resolve Self type references in delegate method assist

rmehri01 opened the pull request:
This PR makes the delegate method assist resolve any `Self` type references in the parameters or return type. It also works across macros such as the `uint_impl!` macro used for `saturating_mul` in the issue example.

Closes #14485

Veykril commented:
Looking at the assist, I'm surprised we aren't using the `PathTransform` yet to fix up paths. We definitely should do that here instead (and then add support for substituting `Self` appropriately). Not sure how simple that is though.

rmehri01 commented:
Yeah that sounds like a good idea, I could give it a try! `PathTransform` expects a `SemanticsScope` for the target scope though, so is there a way to get that with a macro? Like in this example, since `ctx.sema.scope` won't work:

```rust
macro_rules! test_method {
    () => {
        fn test(self) -> Self {
            self
        }
    };
}

struct Bar;

impl Bar {
    test_method!();
}

struct Foo {
    bar: $0Bar,
}
```

Veykril commented:
Why wouldn't it work there? We can take a scope from a macro expansion's node just fine

rmehri01 commented:
Ah I was running into an issue where it said the lookup failed in this instance of `Semantics` but I added `sema.parse_or_expand`, which seems to update the cache and gets it to work! 

I added the change to `PathTransform` to change `Self` now but I'm not sure if its the right way to do it, seems to at least pass the tests I had before.

Veykril commented on crates/ide-assists/src/handlers/generate_delegate_methods.rs:
```diff
@@ -106,7 +107,10 @@ pub(crate) fn generate_delegate_methods(acc: &mut Assists, ctx: &AssistContext<'
             |edit| {
                 // Create the function
                 let method_source = match method.source(ctx.db()) {
```
```suggestion
                let method_source = match ctx.sema.source(method) {
```
We can do this instead of the extra `parse_or_expand`, this will do the caching for us then

Veykril commented on crates/ide-assists/src/handlers/generate_delegate_methods.rs:
```diff
@@ -183,6 +187,12 @@ pub(crate) fn generate_delegate_methods(acc: &mut Assists, ctx: &AssistContext<'
                 let assoc_items = impl_def.get_or_create_assoc_item_list();
                 assoc_items.add_item(f.clone().into());
 
+                PathTransform::generic_transformation(
+                    &ctx.sema.scope(strukt.syntax()).unwrap(),
+                    &ctx.sema.scope(method_source.syntax()).unwrap(),
+                )
+                .apply(f.syntax());
```
```suggestion
                if let Some((target, source)) =
                    ctx.sema.scope(strukt.syntax()).zip(ctx.sema.scope(method_source.syntax()))
                {
                    PathTransform::generic_transformation(&target, &source).apply(f.syntax());
                }
```
The unwraps shouldn't really be possible to hit here, but we are trying to not use unwraps in assist code if possible.

rmehri01 replied:
Ah okay, sorry about that!

Veykril commented on crates/ide-db/src/path_transform.rs:
```diff
@@ -327,8 +327,36 @@ impl Ctx<'_> {
+                    }
+                }
+
+                ted::replace(path.syntax(), ast_ty.syntax());
```
We probably wanna prefer keeping `Self` if it does not need to be replaced. So a check if the source and target scopes have the same `Self` (and in that case, skip this transformation here entirely). We can give `Ctx` a new bool field for that and set it accordingly in `build_ctx` by giving `SemanticsScope` a function like `fn has_same_self_type(&self, other: &SemanticsScope<'_>) -> bool` which just checks if the top `ImplDefScope`s of each scope's resolver has the same def in them.

Veykril commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit d8f0a9dd3bd204571e34a4d59ce3aa2e46069cbd has been approved by `Veykril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust-analyzer) for this repository.

<!-- @bors r=Veykril d8f0a9dd3bd204571e34a4d59ce3aa2e46069cbd -->
<!-- homu: {"type":"Approved","sha":"d8f0a9dd3bd204571e34a4d59ce3aa2e46069cbd","approver":"Veykril","queue":"https://bors.rust-lang.org/homu/queue/rust-analyzer"} -->

bors commented:
:hourglass: Testing commit d8f0a9dd3bd204571e34a4d59ce3aa2e46069cbd with merge fbe39bba3071b42b39077bdfb47b2d1d9b183926...
<!-- homu: {"type":"BuildStarted","head_sha":"d8f0a9dd3bd204571e34a4d59ce3aa2e46069cbd","merge_sha":"fbe39bba3071b42b39077bdfb47b2d1d9b183926"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust-analyzer/actions/runs/7140654301/job/19446443606)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-analyzer/actions/runs/7140654301/job/19446443606","builder_name":"checks-actions"} -->

Veykril commented:
rebased
@bors r+

bors commented:
:pushpin: Commit 7e768cbe703188bb8c126d1be6960637795e4943 has been approved by `Veykril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust-analyzer) for this repository.

<!-- @bors r=Veykril 7e768cbe703188bb8c126d1be6960637795e4943 -->
<!-- homu: {"type":"Approved","sha":"7e768cbe703188bb8c126d1be6960637795e4943","approver":"Veykril","queue":"https://bors.rust-lang.org/homu/queue/rust-analyzer"} -->

bors commented:
:hourglass: Testing commit 7e768cbe703188bb8c126d1be6960637795e4943 with merge 6bbb2ac3049d508b7e03d3968100ed96ca041fb6...
<!-- homu: {"type":"BuildStarted","head_sha":"7e768cbe703188bb8c126d1be6960637795e4943","merge_sha":"6bbb2ac3049d508b7e03d3968100ed96ca041fb6"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust-analyzer/actions/runs/7140815581/job/19447080276)
Approved by: Veykril
Pushing 6bbb2ac3049d508b7e03d3968100ed96ca041fb6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Veykril","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust-analyzer/actions/runs/7140815581/job/19447080276"},"merge_sha":"6bbb2ac3049d508b7e03d3968100ed96ca041fb6"} -->
